### PR TITLE
[Model] Support MiniCPM-RobotTrack

### PR DESCRIPTION
**Pull request by tc-mb**
## Purpose

Add vLLM support for MiniCPM-RobotTrack, a compact vision-language-action tracker built on MiniCPM4-0.5B for embodied target tracking. It combines natural-language instructions with fused DINOv3 and SigLIP visual features to directly predict eight future [x, y, yaw] waypoints for embodied person following.

This is the first version of the mInicpm series target tracking model.

This adds tracker registration, trajectory pooling, weight loading, tests, and documentation.

### CODE REVIEW

File: vllm/model_executor/models/minicpm_robottrack.py [hunk 1]
```diff
@@ -0,0 +1,1201 @@
...
+
+    def _get_image_processor(self, attr: str, loader: Any, model: str) -> Any:
+        cached = getattr(self, attr, None)
+        if cached is None:
+            cached = loader.from_pretrained(model)
+            setattr(self, attr, cached)
+        return cached
+
+    def get_dino_processor(self) -> Any:
+        from transformers import AutoImageProcessor
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpm_robottrack.py]: Please move the imports to top level

Comment (oreomaker) [reply vllm/model_executor/models/minicpm_robottrack.py]: fixed it

File: vllm/model_executor/models/minicpm_robottrack.py [hunk 2]
```diff
@@ -0,0 +1,1199 @@
...
+            quant_config,
+            require_post_norm=True,
+            use_head=False,
+            prefix="siglip",
+        )
+        self.num_register = int(getattr(dino_config, "num_register_tokens", 0) or 0)
+
+        self._load_encoder_weights(dino_model, siglip_model)
+        self.eval()
+        self.requires_grad_(False)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpm_robottrack.py]: Also remove the training specific code

File: vllm/model_executor/models/minicpm_robottrack.py [hunk 3]
```diff
@@ -0,0 +1,1199 @@
...
+
+    @property
+    def dtype(self) -> torch.dtype:
+        return next(self.parameters()).dtype
+
+    @property
+    def device(self) -> torch.device:
+        return next(self.parameters()).device
+
+    @torch.inference_mode()
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpm_robottrack.py]: `torch.inference_mode` is redundant as vLLM does that already

File: MiniCPM-RobotTrack-Encoder-vLLM-IMPLEMENTATION.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
+# MiniCPM-RobotTrack 编码器 → vLLM 集成实现报告(初步版本)
```
Comment (DarkLight1337) [inline MiniCPM-RobotTrack-Encoder-vLLM-IMPLEMENTATION.md]: Remember to remove the agent docs before the final version of the PR

**Comment by oreomaker** [comment]
Current implementation predict 8 `[x, y, yaw]` velocity waypoints (non-generative) based on a 32 frame sliding window encoded by DINOv3 + SigLIP tower. The last frame is the current frame (fine, 8×8 adaptive pool → 64 tokens); the rest are history (coarse, 2×2 pool → 4 tokens/frame); if fewer than 31 history frames, the oldest is repeated to pad to 31. The query is composed as: `[history 31×4 + 31 markers] + [current 64 + 1 marker] + [1 control]`. The whole window is re-encoded every step.

There's a question that needed a discussion with the community -- should we  add the encode cache for MiniCPM-RobotTrack? Several concerns:
* Is the cache worth adding? DINOv3-S + SigLIP are not large, it is OK without the encode cache.
* Going through vLLM's mm cache is feasible but non-trivial. RobotTrack's structured temporal sequence does not fit vLLM's standard mm caching model. It requires reworking the processor's item split and placeholder bookkeeping.

 I'd like the community's view.

**Comment by DarkLight1337** [comment]
I think latency is quite important for the main use case, so we should try to include the cache unless you really think that it's good enough already.

Regarding the caching, perhaps we could add metadata to the input images so that the hash for previous and current frames becomes different, even if the visual content is the same. cc @Isotr0py

File: vllm/model_executor/models/minicpm_robottrack.py (hunk 2 above)
Comment (oreomaker) [reply vllm/model_executor/models/minicpm_robottrack.py]: done

File: vllm/model_executor/models/minicpm_robottrack.py (hunk 3 above)
Comment (oreomaker) [reply vllm/model_executor/models/minicpm_robottrack.py]: done

**Comment by Isotr0py** [comment]
> Regarding the caching, perhaps we could add metadata to the input images so that the hash for previous and current frames becomes different, even if the visual content is the same.

I think we can append some metadata into exif through data parser? Then we can reuse the existing exif hashing code path.
https://github.com/vllm-project/vllm/blob/fbb1ef68030991a14291231bf8877063b57e1ada/vllm/multimodal/hasher.py#L61-L75

File: vllm/model_executor/models/minicpm_robottrack.py [hunk 5]
```diff
@@ -0,0 +1,1196 @@
...
+            quant_config=quant_config,
+            prefix=f"{prefix}.k_proj",
+        )
+        self.v_proj = ColumnParallelLinear(
+            self.embed_dim,
+            self.embed_dim,
+            bias=config.value_bias,
+            quant_config=quant_config,
+            prefix=f"{prefix}.v_proj",
+        )
```
Comment (Isotr0py) [inline vllm/model_executor/models/minicpm_robottrack.py]: Let's use `QKVParallelLinear` to fuse qkv_proj

Comment (oreomaker) [reply vllm/model_executor/models/minicpm_robottrack.py]: DINOv3 uses query_bias=true and key_bias=false, so the QKVParallelLinear may not suitable for this.

Comment (Isotr0py) [reply vllm/model_executor/models/minicpm_robottrack.py]: You can refer to whisper to initialize key bias with full zeros:
https://github.com/vllm-project/vllm/blob/e7b8d5946095a594af2cf7ca3c314b9806cb7c32/vllm/model_executor/models/whisper.py#L1000-L1012

Comment (oreomaker) [reply vllm/model_executor/models/minicpm_robottrack.py]: Added at commit ad67b27

**Comment by oreomaker** [comment]
> > Regarding the caching, perhaps we could add metadata to the input images so that the hash for previous and current frames becomes different, even if the visual content is the same.
> 
> I think we can append some metadata into exif through data parser? Then we can reuse the existing exif hashing code path.
> 
> https://github.com/vllm-project/vllm/blob/fbb1ef68030991a14291231bf8877063b57e1ada/vllm/multimodal/hasher.py#L61-L75

# Frame Cache and Stateful Streaming Window Discussion

Commit 8234c4f adds two bounded, content-addressed LRU caches, both keyed by a per-frame content hash (blake2b of the frame bytes): a GPU-side feature cache that stores each frame's pooled coarse features, and a CPU-side pixel cache that stores resized + normalized pixels. Each step, only the frames the hash misses — normally just the newly arrived one — are encoded or normalized; the unchanged history frames are served from cache.

Measured on an RTX 4090 (50 frames, fp16, eager), steady-state ms/step:

cache | ms/step
--- | ---
off | 524
frame cache  | 408
frame + pixel cache | 214

A large cost remains, though: every step still moves ~114 MB of normalized pixels across the client→engine IPC boundary, even though 31 of 32 frames are identical to the previous step. That transport accounts for roughly 118 ms of the ~214 ms per-step cost.

## Proposed next step: a stateful stream protocol (working tree)

Move the window state into the server; the client sends one new frame per step:

```
establish: {"frames": [32], "stream_id": "cam0", "frame_index": 31}   # replace
advance:   {"frames": [1],  "stream_id": "cam0", "frame_index": 32}   # append
retry:     {"frames": [1],  "stream_id": "cam0", "frame_index": 32}   # reuse (idempotent)
```

Server side, per stream: 31-frame coarse history + current fine + last frame_index, in an LRU (`max_cached_streams`, default 8) on the model, keyed by a hash of `stream_id`. A full window replaces; a consecutive single frame appends; a repeated `frame_index` is an idempotent retry served from committed state; anything else is an error (unestablished / out-of-order / evicted).

On the same 4090: steady ~57 ms/step vs ~214 ms stateless, stateful vs stateless agree to max|Δ|=1.8e-4. All model-side, no vLLM core changes (stream_id/frame_index ride through as batched mm fields).

The trade-off is: the server now holds session state, DP>1 needs session affinity or DP=1 (considering the model is small, it should be reasonable).

For reviewers: is the model side session state the right direction for vLLM? Thanks for any feedback.
